### PR TITLE
provider: add input_type pem for RSA decoder

### PR DESCRIPTION
**Pull request by danbev**
This commmit adds the input_type pem to the RSA decoder.

The motivation for this is an issue we ran into when upgrading Node.js
to OpenSSL 3.0. When reading a BIO, there will be a peek of the file to
see if it is of type PEM:
```c
void *file_attach(void *provctx, OSSL_CORE_BIO *cin)
{
  ...
  peekbuf[sizeof(peekbuf) - 1] = '\0';
  if (strstr(peekbuf, "-----BEGIN ") != NULL)
    input_type = INPUT_TYPE_PEM;
  ...
  if (BIO_tell(new_bio) != loc) {
    /* In this case, anything goes */
    input_type = INPUT_TYPE_ANY;
}
```
When calling PEM_read_bio_PrivateKey with a BIO from a file, BIO_tell
would be supported and the input type is set to INPUT_TYPE_ANY (so we
don't hit this issue in that case). But in Node's case the input_type
will be "pem" and later when trying to find a decoder for that
input_type leading to the issue we are seeing.

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
It would be nice if you could refer to the issue you're talking about.
The basic issue here is a fragility in the decoder, that it essentially requires BIO_tell() support. I thought I had fended that off in the attach function, but it sounds like not.

The fix provided here is a bit of a hack, rather than a complete solution.

**Comment by levitte** [comment]
How is the input from Node.js given, and from what kind of source? We probably need a test case that simulates that.

**Comment by danbev** [comment]
@levitte I've got some details about this issue [here](https://github.com/danbev/learning-libcrypto#rsa-pem-decoder-issue). This issue was only discovered after https://github.com/openssl/openssl/pull/12901.

For Node.js we have this tracking issue: https://github.com/nodejs/node/issues/29817

> How is the input from Node.js given, and from what kind of source? We probably need a test case that simulates that.

Node uses something called [NodeBIO](https://github.com/nodejs/node/blob/master/src/node_crypto_bio.h). I'd be happy to add a test that mimics what Node does.

**Comment by levitte** [comment]
Ok...  I just read what you've written in [your write-up](https://github.com/danbev/learning-libcrypto#rsa-pem-decoder-issue), and I detect one misunderstanding:

> Now, in the case of Node.js there is a in memory BIOs that does not support BIO_tell, but when reading from a file BIO_tell would be supported and the input type is set to null (so we don't hit this issue in that case). In our case the input_type will be "pem". *But later when trying to find a decoder for that input type there will not be one available.*

There's a fine point about the new decoders, and it's that they can form a chain.  So while there is no direct PEM -> RSA decoder, there are a PEM -> DER decoder and a DER -> RSA decoder, so the decoding ends up happening in two steps, essentially doing PEM -> DER -> RSA.  We did it this way to allow other providers to supply decoders from other on-file formats to DER if they would like to.

In your analysis of what happens, you're absolutely right that the call to `ossl_decoder_ctx_setup_for_EVP_PKEY` does build up a stack of decoders to try when running the decoding process to get a suitable key...  from DER.  The `OSSL_DECODER_CTX_add_extra` call that follows does the rest, i.e. it adds further decoders that do format translation, thus building the possible chains.

That being said:

> I'd be happy to add a test that mimics what Node does.

Please do.  I very strongly suspect that this is a case of bad interaction between the buffer BIO that `file_attach` adds to be able to peek, and that it's either something odd going on with BIO_tell / BIO_seek, or the peeking reads too muich...  I'd be very interested to see if that hypothesis holds some truth.

**Comment by danbev** [comment]
@levitte I've tried to mimic what the node.js BIO does in the latest commit. I also cherry-picked https://github.com/openssl/openssl/pull/12901 as otherwise we would not get any further than that error. I also removed the change to providers.inc to that this test fails which hopefully makes it easier to debug.

Hopefully you should be able to build and then run the test using:
```console
$ ./test/pemtest -test test_PEM_read_bio_PrivateKey_with_in_mem_nodejs_bio
```

File: crypto/store/store_lib.c [hunk 1]
```diff
@@ -932,9 +932,16 @@ OSSL_STORE_CTX *OSSL_STORE_attach(BIO *bp, const char *scheme,
...
+    ERR_set_mark();
     if ((loader = ossl_store_get0_loader_int(scheme)) != NULL)
         loader_ctx = loader->attach(loader, bp, libctx, propq,
                                     ui_method, ui_data);
+    /*
+     * ossl_store_get0_loader_int will will raise an error if the loader for
+     * the scheme cannot be retrieved. This error is cleared as the code below
+     * will try to fetch the loader from the provider.
+     */
+    ERR_pop_to_mark();
```
Comment (levitte) [inline crypto/store/store_lib.c:944]: Ah!  Good thinking, but perhaps a little be coarse.  Have a look at `OSSL_STORE_open_with_libctx`, which does exactly this, but slightly differently to allow these errors to come through if there was no loader available at all.

Thank you, not having this error handling at all in here is an omission and does need fixing.

File: test/endecode_test.c [hunk 2]
```diff
@@ -1023,6 +1024,27 @@ static int create_ec_explicit_trinomial_params(OSSL_PARAM_BLD *bld)
...
+  ret = TEST_true(ossl_decoder_ctx_setup_for_EVP_PKEY(dctx, &pkey, ctx, NULL))
+        && TEST_ptr(pem_decoder = OSSL_DECODER_fetch(ctx, "RSA",
+        "provider=default,fips=yes,input=pem"));
+
+  OSSL_DECODER_free(pem_decoder);
+  OSSL_DECODER_CTX_free(dctx);
+  OPENSSL_CTX_free(ctx);
+  return ret;
+}
+
```
Comment (levitte) [inline test/endecode_test.c:1047]: This test shouldn't be necessary.

File: crypto/store/store_lib.c (hunk 1 above)
Comment (danbev) [reply crypto/store/store_lib.c:944]: Thanks, I'll take a closer look at how this is done in `OSSL_STORE_open_with_libctx`. This particular change is part of https://github.com/openssl/openssl/pull/12901 so I'll make the changes to that PR and in this PR. Sorry if this was confusing but without including that change the issue in this PR would not be noticed.

File: test/endecode_test.c (hunk 2 above)
Comment (danbev) [reply test/endecode_test.c:1047]: Thanks, I'll remove it.

File: crypto/store/store_lib.c (hunk 1 above)
Comment (levitte) [reply crypto/store/store_lib.c:944]: No problem whatsoever, thanks for informing me.

**Comment by danbev** [comment]
Just a note that the test in https://github.com/openssl/openssl/pull/12971/commits/7a9c287a49840e595f3246e558e7f2c5906f5da2 is expected to fail and was added to help simulate Node.js BIO.

**Comment by danbev** [comment]
Closing as this is out of date.

**Pull request closed without merging**